### PR TITLE
Require Mesh::addCoordinates(location) to initialize Coordinates at location

### PR DESCRIPTION
**Pull request by johnomotani**
This is the 'final' version of changes started in #1372. Now only CELL_CENTRE coordinates are created by default, and the user must call REQUEST_LOCATION(location) or mesh->addCoordinates(location) to initialize Coordinates for staggered locations.

Includes updates to the manual, unit tests for `addCoordinates()` as part of `BoutMeshTest` and support for `getCoordinates`/`addCoordinates` in boutcore.

This does add a couple of methods to `Mesh`, `hasCoordinates(location)` and `countCoordinates()`. I think they're justified, but alternatives could be to make `Coordinates` a friend of `Mesh`, or to make coords_map a public member; I don't favour either of those alternatives, but maybe there are others?

### CODE REVIEW

File: src/mesh/mesh.cxx [hunk 1]
```diff
@@ -325,13 +326,47 @@ ParallelTransform& Mesh::getParallelTransform() {
...
+        // location does already exist in coords_map, so reset it
+
+        // first erase the existing entry to avoid throwing an exception from
+        // Coordinates::geometry().
+        // The check that would throw the exception is not needed because
+        // replacement of the Coordinates object has been explicitly requested.
+        coords_map.erase(location);
+        coords_map.emplace(location, std::unique_ptr<Coordinates>(new Coordinates(this, location, getCoordinates(CELL_CENTRE))));
+      } else {
+        throw BoutException("Coordinates at %s already added to Mesh",
```
Comment (ZedThree) [inline src/mesh/mesh.cxx]: What about:

```cpp
if (!replace_coords and (coords_map.count(location) > 0)) {
  throw BoutExcept(...);
}
if (replace_coords) {
  coords_map.erase(location);
}
coords_map.emplace(location,
                   std::unique_ptr<Coordinates>(new Coordinates(this, location, getCoordinates(CELL_CENTRE))))
```

- precondition up front
- removes duplicated code
- interesting bit not in conditional

---

Also, here is where `std::make_unique` would be very useful, but it is unfortunately C++14. I can add a replacement that we can use until we upgrade

File: tools/pylib/_boutcore_build/boutcore.pyx.in [hunk 2]
```diff
@@ -740,14 +738,30 @@ cdef class Mesh:
...
+        loc_ = benum.resolve_cell_loc(location)
+        return coordsFromObj(self.cobj.getCoordinates(loc_))
+
     @property
     def coordinates(self):
         """
-        Get the Coordinates object of this mesh
+        Deprecated version of getCoordinates
+        """
+        print("Warning Mesh.coordinates is deprecated, and does not handle "
```
Comment (ZedThree) [inline tools/pylib/_boutcore_build/boutcore.pyx.in]: Maybe use `warnings.warn`?

File: include/bout/mesh.hxx [hunk 3]
```diff
@@ -434,17 +435,35 @@ class Mesh {
...
+    return coords_map.count(location);
+  }
+
+  /// Count how many Coordinates objects have been added
+  int countCoordinates() {
+    return coords_map.size();
+  }
+
+  /// switch to pass to Coordinates objects
+  bool allow_geometry_without_recalculate_staggered;
```
Comment (ZedThree) [inline include/bout/mesh.hxx]: Please could the doxygen comment be a bit more descriptive?

Also should this be a member of `mesh`/does it need to be stored at all? It looks like it should maybe just be an argument to `Coordinates::geometry`?

**Comment by ZedThree** [comment]
I'm afraid I'm not 100% sure about the utility of the macro. It's not much shorter than `mesh->addCoordinates`.

Would an alternative to requiring users to call this in their physics models be to put the call to `Mesh::addCoordinates` in `Field3D::applyBoundary`?

**Comment by johnomotani** [comment]
I'd be fine with getting rid of the macro if you think it's not useful.

Why in `Field3D::applyBoundary`? I'd thought about trying to put the call to `Mesh::addCoordinates` in `Field*::setLocation` but that gets called frequently (it's inside a whole bunch of other methods). I'm in favour of requiring the user to make an explicit call to something because the order of calling for example `Mesh::addCoordinates` and `Coordinates::geometry` matters; making the user call `Mesh::addCoordinates` seems like the only way to have `addCoordinates` be called once and only once for each location, which I think is attractive because it avoids recalculating stuff and minimizes the amount of checking needed.

**Comment by johnomotani** [comment]
Fabio just made a good point: the way this PR has gone, we're now initializing settings for staggered grids in two ways - with the `mesh:staggergrids` option in the input file and with `Mesh::addCoordinates` c++-code in the physics model `init()` [even before, some 'settings' are in c++-code since the user has to call `setLocation` or use `outloc` arguments to derivatives, etc.].

This might be a good time to simplify the initialization. We could either:
1. Have input file settings specifying which locations should be available, e.g. `mesh:staggered locations = xlow, ylow, zlow`. `Coordinates` objects get created for all locations selected, and updated whenever `geometry()` is called on the `CELL_CENTRE` `Coordinates`. Remove `Mesh::addCoordinates`, or make it protected.
2. Remove 'staggering' settings from the input file and just require a call to `Mesh::addCoordinates` during `init` for each location that will be used. `addCoordinates` should be called after `geometry()` on the `CELL_CENTRE` `Coordinates` and before any `setLocation`. Since models like STORM that use staggering probably hard-code the `CELL_LOC`s anyway, removing run-time options for staggering could be sensible.

In either case I think the `mesh:staggergrids` option and `bool Mesh::StaggerGrids` could be removed/deprecated now. `Mesh::StaggerGrids` is actually not used much any more except in checks where we wouldn't lose anything by leaving the check as something like `inloc==outloc` or `rhs.getLocation()==lhs.getLocation()`. I think the only place it makes any real difference is in `derivs_initialise()` where staggered derivative tables aren't created, but #1384 replaces that anyway.

Thoughts? I'm inclined to option 2, a compile-time 'setting' in `init()`.

**Comment by dschwoerer** [comment]
I would also prefer option 2.
It can still be turned into a run-time switch, if that is of interest.
The user code anyway need to check whether staggering is enabled, before
trying to set a staggered location.

File: include/bout/mesh.hxx [hunk 4]
```diff
@@ -708,7 +727,7 @@ class Mesh {
   
   GridDataSource *source; ///< Source for grid data
   
-  std::map<CELL_LOC, std::shared_ptr<Coordinates> > coords_map; ///< Coordinate systems at different CELL_LOCs
+  std::map<CELL_LOC, std::unique_ptr<Coordinates> > coords_map; ///< Coordinate systems at different CELL_LOCs
```
Comment (ZedThree) [inline include/bout/mesh.hxx]: I'm not sure about changing this to `unique_ptr`. I know we currently hand out a raw pointer, but that should change, and returning a `unique_ptr` says "I'm giving ownership over to you", whereas a `shared_ptr` says "You can have a look at this, but I will keep hold of it too"

**Comment by bendudson** [comment]
I think if the result can't be null,  and we're not transferring or sharing ownership,  then a reference would be the right way to go. I may have got that wrong though.

**Comment by johnomotani** [comment]
If in the long term we want to return a reference, what should we do now? Changing `getCoordinates()` to return a reference would break backward compatibilty. Do we want a new method that returns a reference (what would be a sensible name for it?) and to deprecate `Coordinates* getCoordinates()`? Or to leave as-is (returning a pointer) for now, and change to returning a reference in the next major release? Or something else?

**Comment by bendudson** [comment]
Since coordinates only recently changed to getCoordinates, perhaps getCoordinates could return a reference,  and coordinates return a pointer?

**Comment by johnomotani** [comment]
> Since coordinates only recently changed to getCoordinates, perhaps getCoordinates could return a reference, and coordinates return a pointer?

I'd be happy with that, but technically it changes the v4.2 interface... Any thoughts @ZedThree?

**Comment by johnomotani** [comment]
Interface question: Rather than requiring users to call `Mesh::addCoordinates(location)` for each `location` they want to use, would it be better to have a separate method like `Mesh::enableLocation(location)`? For now `enableLocation(location)` could just call `addCoordinates(location)` but it would be more future-proof in case we need to do other things for different locations, for example create different `ParallelTransform`s, at some point.

**Comment by ZedThree** [comment]
I think we're passing this up in favour of `setLocation` automagically doing the right thing (reopen this if I'm wrong!)

**Pull request closed without merging**